### PR TITLE
feat: Support CartesianProductExec in comet

### LINKED ISSUE #199: Support CartesianProductExec
### What is the problem the feature request solves?

Spark uses `CartesianProductExec` for cross join operation. This ticket is for the support of `CartesianProductExec` in Comet.

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leoluan2009**
## Which issue does this PR close?

Closes #199 .
TODO: need to support join condition after CrossJoin in Datafusion support join condition

### CODE REVIEW

**Comment by leoluan2009** [comment]
@viirya help to start CI and review, thanks

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala [hunk 1]
```diff
@@ -899,6 +899,39 @@ case class CometSortMergeJoinExec(
...
+        false
+    }
+  }
+
+  override def hashCode(): Int = Objects.hashCode(condition, left, right)
+
+  override lazy val metrics: Map[String, SQLMetric] =
+    CometMetricNode.baselineMetrics(sparkContext)
+}
 case class CometScanWrapper(override val nativeOp: Operator, override val originalPlan: SparkPlan)
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:935]:
```suggestion
}

case class CometScanWrapper(override val nativeOp: Operator, override val originalPlan: SparkPlan)
```

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 2]
```diff
@@ -2545,6 +2545,18 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
         withInfo(join, "SortMergeJoin is not enabled")
         None
 
+      case join: CartesianProductExec if isCometOperatorEnabled(op.conf, "cross_join") =>
+        // TODO: Support CartesianProductExec with join condition after new DataFusion release
+        if (join.condition.isDefined) {
+          withInfo(op, "cross_join with a join condition is not supported")
+          return None
+        }
+        None
```
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2554]: Seems you don't create the protobuf object for CrossJoin?

Comment (leoluan2009) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: it is my mistake

**Comment by leoluan2009** [comment]
@viirya thanks for your review, resolve your commets

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 3]
```diff
@@ -698,6 +698,31 @@ class CometSparkSessionExtensions
...
+                op.condition,
+                op.left,
+                op.right,
+                SerializedPlan(None))
+            case None =>
+              op
+          }
+
+        case op: CartesianProductExec if !isCometOperatorEnabled(conf, "cross_join") =>
+          withInfo(op, "Cross_Join is not enabled")
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:719]: nit: `Cross_Join` => `CrossJoin`

**Review by kazuyukitanimura — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
@leoluan2009 You may need to generate plan stability results to pass the CI pipelines.

File: spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala [hunk 4]
```diff
@@ -251,4 +251,18 @@ class CometJoinSuite extends CometTestBase {
       }
     }
   }
+
+  // TODO: Add a test for CartesianProductExec with join filter after new DataFusion release
```
Comment (parthchandra) [inline spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala:255]: Add a github issue for the followup and refer to it here?

**Comment by leoluan2009** [comment]
> @leoluan2009 You may need to generate plan stability results to pass the CI pipelines.

OK, I will update soon, thanks!

File: spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala (hunk 4 above)
Comment (leoluan2009) [reply spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala:255]: fine

**Comment by viirya** [comment]
@leoluan2009 Would you like to update the plan stability results?

**Comment by leoluan2009** [comment]
> @leoluan2009 Would you like to update the plan stability results? Thanks.

Hi @viirya  sorry for delay reply, I think this pr do not change the query plan for tpcds?

**Comment by viirya** [comment]
> Hi @viirya sorry for delay reply, I think this pr do not change the query plan for tpcds?

Hmm, last time I saw the plans were changed so CI was failed, but now the CI doesn't fail on plan stability.

But CI currently failed on some other tests.

**Pull request closed without merging**